bpf/lib/nat.h

```diff
@@ -382,12 +382,82 @@ static __always_inline int snat_v4_rewrite_egress(struct __ctx_buff *ctx,
 static __always_inline int snat_v4_rewrite_ingress(struct __ctx_buff *ctx,
 						   struct ipv4_ct_tuple *tuple,
 						   struct ipv4_nat_entry *state,
-						   __u32 off)
+						   __u32 off, __u8 nexthdr)
 {
 	int ret, flags = BPF_F_PSEUDO_HDR;
 	struct csum_offset csum = {};
 	__be32 sum_l4 = 0, sum;
 
+	if (state->to_daddr == tuple->daddr &&
+	    state->to_dport == tuple->dport)
+		return 0;
+	sum = csum_diff(&tuple->daddr, 4, &state->to_daddr, 4, 0);
+	csum_l4_offset_and_flags(nexthdr, &csum);
+	if (state->to_dport != tuple->dport) {
+		switch (nexthdr) {
+		case IPPROTO_TCP:
+		case IPPROTO_UDP:
```

Comment:
@sahid Thanks for your work on this!

I'll admit that I've lost track of this PR, but it's not clear to me how the changes here relate to handling`ICMP_FRAG_NEEDED`. Am I missing something?

Might be worth adding a comment and/or some text in the commit message to describe what these changes are and why  they are needed.

FWIW, I would suggest to not do any code changes yet. Wait for feedback from other reviewers first, since CI is now green.

Reply:
Thanks a lot @kkourt to have took time looking at this IPv4 implementation I have a the IPv6 ready based on that one.

About your ask I have created a method `rewrite_embedded` which changes the source IP of the embedded packet and also the for TCP/UDP the port if it has been changed during the egrees NAT, for ICMP the ID.

The method is quite similar to `rewrite_ingress`, so the diff completely mixed-up the thing which is making the review very difficult.I will try in my next update to move the method, so I would expect the diff to avoid that mix. 

I will also add more comments. That makes lot of sense yes.

As I mentioned I have the IPv6 ready but I will wait to get more comments, probably from @brb :-)